crates/context_aware_config/src/api/dimension/handlers.rs

```diff
@@ -50,76 +52,175 @@ async fn create(
-            ));
-        }
-    };
+    validate_dimension_position(
+        create_req.dimension.clone(),
+        create_req.position.clone(),
+        num_rows,
+    )?;
+    validate_jsonschema(&state.meta_schema, &schema_value)?;
 
-    let new_dimension = Dimension {
+    let dimension_data = Dimension {
         dimension: create_req.dimension.into(),
-        priority: create_req.priority.into(),
-        position: 0, // hard coded for now till we migrate
+        position: create_req.position.into(),
         schema: schema_value,
         created_by: user.get_email(),
         created_at: Utc::now(),
-        function_name: fun_name.clone(),
+        function_name: create_req.function_name.clone(),
         last_modified_at: Utc::now().naive_utc(),
         last_modified_by: user.get_email(),
     };
 
-    let upsert = diesel::insert_into(dimensions)
-        .values(&new_dimension)
-        .on_conflict(dimensions::dimension)
-        .do_update()
-        .set(&new_dimension)
-        .get_result::<Dimension>(&mut conn);
-
-    match upsert {
-        Ok(upserted_dimension) => {
-            let is_mandatory = tenant_config
-                .mandatory_dimensions
-                .contains(&upserted_dimension.dimension);
-            Ok(HttpResponse::Created().json(DimensionWithMandatory::new(
-                upserted_dimension,
-                is_mandatory,
-            )))
-        }
-        Err(diesel::result::Error::DatabaseError(
-            diesel::result::DatabaseErrorKind::ForeignKeyViolation,
-            e,
-        )) => {
-            log::error!("{fun_name:?} function not found with error: {e:?}");
-            Err(bad_argument!(
-                "Function {} doesn't exists",
-                fun_name.unwrap_or(String::new())
-            ))
-        }
-        Err(e) => {
-            log::error!("Dimension upsert failed with error: {e}");
-            Err(unexpected_error!(
-                "Something went wrong, failed to create/update dimension"
+    conn.transaction::<_, superposition::AppError, _>(|transaction_conn| {
+        diesel::update(dimensions)
+            .filter(dimensions::position.ge(dimension_data.position))
+            .set((
+                last_modified_at.eq(Utc::now().naive_utc()),
+                last_modified_by.eq(user.get_email()),
+                dimensions::position.eq(dimensions::position + 1),
             ))
+            .execute(transaction_conn)?;
+        let insert_resp = diesel::insert_into(dimensions)
+            .values(&dimension_data)
+            .get_result::<Dimension>(transaction_conn);
+
+        match insert_resp {
+            Ok(inserted_dimension) => {
+                let is_mandatory = tenant_config
+                    .mandatory_dimensions
+                    .contains(&inserted_dimension.dimension);
+                Ok(HttpResponse::Created().json(DimensionWithMandatory::new(
+                    inserted_dimension,
```

Comment:
nab but curious, if someone changes the tenant config to remove mandatory status for a dimension do we set false here @mahatoankitkumar ? If not we can add it in workspace settings

Reply:
@Datron currently from appstate we are injecting in the request and thats being passed on to handler 
we will have to add in workspace_settings and use db calls probably